user_topic_popover: Convert module to TypeScript.

klarabratteby opened the pull request:
<!-- Describe your pull request here.-->

This PR migrates user_topic_popover.js to user_topic_popover.ts.

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

timabbott commented:
Welcome! Can you talk about how you tested this work?

klarabratteby commented:
Hi @timabbott , thank you!!

I ran tools/run-tsc, ./tools/test-all and ./tools/lint --fix   

I also manually tested the UI by opening the popover in the web interface to ensure it worked as expected. 

Please let me know if I have missed any steps!

alya commented:
@timabbott are you reviewing this PR, or would you like someone else to take a look first?

klarabratteby commented:
Hi @timabbott and @alya,
Are there any additional steps required before this PR can move forward for review or merging? 😊

LaPulgaaa commented on web/src/user_topic_popover.ts:
```diff
@@ -54,11 +64,11 @@ export function initialize() {
 
                 const visibility_policy = Number.parseInt(
-                    $(e.currentTarget).attr("data-visibility-policy"),
+                    $(e.currentTarget).attr("data-visibility-policy") ?? "0",
```
We can null assert (`!`) this. Defaulting this to "0" is a little over defensive.

LaPulgaaa commented on web/src/user_topic_popover.ts:
```diff
@@ -67,10 +77,10 @@ export function initialize() {
                     const prev_visibility_policy = user_topics.get_topic_visibility_policy(
-                        stream_id,
-                        topic_name,
+                        stream_id!,
```
Explicit `assert`s would be more readable.

LaPulgaaa commented on web/src/user_topic_popover.ts:
```diff
@@ -84,8 +94,8 @@ export function initialize() {
                 user_topics.set_user_topic_visibility_policy(
-                    stream_id,
-                    topic_name,
+                    stream_id!,
```
Explicit `assert`s would be more readable.

LaPulgaaa commented on web/src/user_topic_popover.ts:
```diff
@@ -28,24 +32,30 @@ export function initialize() {
+            stream_id = Number.parseInt(stream_id_str, 10);
+            topic_name = topic_name_str;
 
-            instance.context =
```
Removing this hunk can be a prep-commit. You can learn about our commit discipline from [Zulip docs](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html#).

LaPulgaaa commented on web/src/user_topic_popover.ts:
```diff
@@ -28,24 +32,30 @@ export function initialize() {
-            const stream_id = $elt.attr("data-stream-id");
-            const topic_name = $elt.attr("data-topic-name");
+            const stream_id_str = $elt.attr("data-stream-id");
+            const topic_name_str = $elt.attr("data-topic-name");
```
Is the `topic_name_str` variable serving any purpose other than to  just be assigned to a more useful `topic_name` variable four lines downstream? If not than probably, we can just
```
topic_name = $elt.attr("data-topic-name")!;
```

LaPulgaaa reviewed (commented):
Thanks for your contribution @klarabratteby !! I have left some comments that can be addressed before a core contributor review the PR.

LaPulgaaa commented on web/src/user_topic_popover.ts:
```diff
@@ -28,24 +32,30 @@ export function initialize() {
+            stream_id = Number.parseInt(stream_id_str, 10);
+            topic_name = topic_name_str;
 
-            instance.context =
-                popover_menus_data.get_change_visibility_policy_popover_content_context(
-                    Number.parseInt(stream_id, 10),
-                    topic_name,
-                );
```
Removing this hunk can be a prep-commit. You can learn about our commit discipline from [Zulip docs](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html#).

klarabratteby commented:
> Thanks for your contribution @klarabratteby !! I have left some comments that can be addressed before a core contributor review the PR.

Hi @LaPulgaaa , thank you for your valuable feedback!! I’ll review and update the code based on your comments 😊

klarabratteby commented:
Hi @LaPulgaaa , I have addressed your comments now! 😊 While doing the prep-commit I accidentally updated settings_user_topics.ts to settings_user_topics.js in test-js-with-node. However, I have corrected this in the second commit, so everything is as it should be. Please let me know if this is misleading or if there's anything else I need to fix. Thank you!

LaPulgaaa commented on web/src/user_topic_popover.ts:
```diff
@@ -28,19 +32,28 @@ export function initialize() {
 
             instance.setContent(
-                parse_html(render_change_visibility_policy_popover(instance.context)),
+                parse_html(
+                    render_change_visibility_policy_popover(
+                        popover_menus_data.get_change_visibility_policy_popover_content_context(
+                            stream_id,
+                            topic_name,
+                        ),
+                    ),
+                ),
```
This should belong in the fixed commit. We are trying to fix a TypeScript bug ie.`instance` object is being mutated with a field (`context`) that it does not contain. Hence all the occurrences of `instance.context` should be replaced with the correct version in a single prep commit; fixing the problem.

Apologies if I wasn't able to get my point across in the first review.

klarabratteby replied:
@LaPulgaaa I think I understand, but I just want to confirm.

**Prep-commit** (js-file), remove instance.context and replace with:  
instance.setContent(
    parse_html(
        render_change_visibility_policy_popover(
            popover_menus_data.get_change_visibility_policy_popover_content_context(
                Number.parseInt(stream_id, 10),
                topic_name,
            )
        )
    )
);

**Second commit** (ts-file):  
assert(stream_id_str !== undefined);
stream_id = Number.parseInt(stream_id_str, 10);
topic_name = $elt.attr("data-topic-name")!;

            instance.setContent(
                parse_html(
                    render_change_visibility_policy_popover(
                        popover_menus_data.get_change_visibility_policy_popover_content_context(
                            stream_id,
                            topic_name,
                        ),
                    ),
                ),
            );

Does that sound correct?

LaPulgaaa replied:
Yes. LGTM!

LaPulgaaa reviewed (commented):
Thanks for the revision. This mostly looks good except for some nits about the commit standards.

klarabratteby commented:
@LaPulgaaa I’m done with the adjustments now. Tested all automated tests + manually in the UI on both commits, they are working properly from what I could see, both individually and collectively 😊

LaPulgaaa commented:
> @LaPulgaaa I’m done with the adjustments now. Tested all automated tests + manually in the UI on both commits, they are working properly from what I could see, both individually and collectively 😊

Thanks for the review. Looks good to me!!

timabbott commented on web/src/user_topic_popover.ts:
```diff
@@ -8,7 +9,10 @@ import {parse_html} from "./ui_util";
 
-export function initialize() {
+let stream_id: number | undefined;
+let topic_name: string | undefined;
```
Why do we need to introduce these global variables? I feel like the prior code's approach of parsing these values from the DOM was a nicer set of abstractions.

klarabratteby replied:
Hi @timabbott  , thank you so much for your review! I agree with you, they are actually unnecessary... I have removed them now 😊

timabbott commented:
Merged, thanks for the contribution @klarabratteby! Probably wise to pick another similarly sized module for your next PR to practice what you've learned.
